compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -1861,10 +2076,54 @@ fn confirm_impl_candidate<'cx, 'tcx>(
             obligation.cause.span,
             "impl item and trait item have different parameter counts",
         );
-        Progress { ty: err, obligations: nested }
+        Progress { term: err, obligations: nested }
     } else {
         assoc_ty_own_obligations(selcx, obligation, &mut nested);
-        Progress { ty: ty.subst(tcx, substs), obligations: nested }
+        Progress { term: ty.subst(tcx, substs), obligations: nested }
+    }
+}
+
+fn confirm_impl_candidate_const<'cx, 'tcx>(
+    selcx: &mut SelectionContext<'cx, 'tcx>,
+    obligation: &ProjectionTyObligation<'tcx>,
+    impl_impl_source: ImplSourceUserDefinedData<'tcx, PredicateObligation<'tcx>>,
+) -> Progress<'tcx, &'tcx ty::Const<'tcx>> {
+    let tcx = selcx.tcx();
+
+    let ImplSourceUserDefinedData { impl_def_id, substs, nested } = impl_impl_source;
+    let assoc_item_id = obligation.predicate.item_def_id;
+    let trait_def_id = tcx.trait_id_of_impl(impl_def_id).unwrap();
+
+    let param_env = obligation.param_env;
+    let assoc_ct = match assoc_def(selcx, impl_def_id, assoc_item_id) {
+        Ok(assoc_ct) => assoc_ct,
+        Err(ErrorReported) => {
+            return Progress { term: tcx.const_error(tcx.ty_error()), obligations: nested };
+        }
+    };
+
+    if !assoc_ct.item.defaultness.has_value() {
+        return Progress { term: tcx.const_error(tcx.ty_error()), obligations: nested };
+    }
+    let substs = obligation.predicate.substs.rebase_onto(tcx, trait_def_id, substs);
+    let substs =
+        translate_substs(selcx.infcx(), param_env, impl_def_id, substs, assoc_ct.defining_node);
+    let _ty = tcx.type_of(assoc_ct.item.def_id);
+    // TODO need to figure how to get the const of the assoc_ct.item.def_id
+    // I'm not sure if there's another tcx query for it.
+    let _ct = ();
```

Comment:
This is the main part I do not know where to progress, since I have the `DefId` but I am unsure how to get its const. I think it's just the `DefId` of the impl's associated const, but how would I convert that into a `ty::Const`?